# ☕ Coffee Shop Sales: From Business Questions to Data Insights

---

## 📖 Overview
> **"Data is only as good as the questions you ask."**

This notebook is designed as a **Practical Guide** to demonstrate how raw transaction data can be transformed into actionable business intelligence. We move beyond simple coding to focus on **Analytical Storytelling**—answering critical business questions through interactive visualizations.

---

## 🎯 Project Objectives
* **Exploratory Data Analysis (EDA):** Ensuring data integrity, handling formats, and understanding statistical distributions.
* **Business-to-Data Mapping:** Learning how to translate a business need (e.g., "When should we increase staff?") into a technical visualization.
* **Interactive Reporting:** Using **Plotly Express** to build clear, high-impact charts for stakeholders.

---

## 📚 Data Dictionary
Understanding the domain is the first step of any analysis. Here is what our columns represent:

| Column | Description |
| :--- | :--- |
| **`hour_of_day`** | The specific hour (0-23) the sale occurred. |
| **`cash_type`** | Payment method (Card or Cash). |
| **`money`** | The revenue amount generated from the sale. |
| **`coffee_name`** | The type of beverage sold. |
| **`Time_of_Day`** | Categorized periods: Morning, Afternoon, or Night. |
| **`Weekday`** | Day of the week (Mon, Tue, etc.). |
| **`Date`** | The calendar date of the transaction. |

---

## 🛠️ Tech Stack
* **Language:** Python 3.x
* **Libraries:** `Pandas` for data manipulation, `Plotly Express` for interactive charts.

#  Phase 1: Exploratory Data Analysis (EDA)

Before generating insights, we must perform a thorough **EDA** to understand the data's structure and ensure its quality. This phase focuses on:
1. **Data Inspection:** Checking the size and types of our features.
2. **Data Cleaning:** Identifying and handling missing values or duplicates.
3. **Data Transformation:** Converting columns to their appropriate formats (e.g., Dates).
4. **Statistical Profiling:** Summarizing the numerical distribution of sales.

In [1]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots


### 1- Data Inspection & Understanding
Before diving into visualizations, we must understand the structure of our dataset. 
In this section, we load the data and inspect its schema to ensure it's ready for analysis.

In [2]:
# Load the dataset
df=pd.read_csv('Coffe_sales.csv')

# Display basic information about dataframe
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3547 entries, 0 to 3546
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   hour_of_day  3547 non-null   int64  
 1   cash_type    3547 non-null   object 
 2   money        3547 non-null   float64
 3   coffee_name  3547 non-null   object 
 4   Time_of_Day  3547 non-null   object 
 5   Weekday      3547 non-null   object 
 6   Month_name   3547 non-null   object 
 7   Weekdaysort  3547 non-null   int64  
 8   Monthsort    3547 non-null   int64  
 9   Date         3547 non-null   object 
 10  Time         3547 non-null   object 
dtypes: float64(1), int64(3), object(7)
memory usage: 304.9+ KB


,hour_of_day,cash_type,money,coffee_name,Time_of_Day,Weekday,Month_name,Weekdaysort,Monthsort,Date,Time
0,10,card,38.7,Latte,Morning,Fri,Mar,5,3,2024-03-01,10:15:50.520000
1,12,card,38.7,Hot Chocolate,Afternoon,Fri,Mar,5,3,2024-03-01,12:19:22.539000
2,12,card,38.7,Hot Chocolate,Afternoon,Fri,Mar,5,3,2024-03-01,12:20:18.089000
3,13,card,28.9,Americano,Afternoon,Fri,Mar,5,3,2024-03-01,13:46:33.006000
4,13,card,38.7,Latte,Afternoon,Fri,Mar,5,3,2024-03-01,13:48:14.626000


> ####  What did we learn from `.info()`?
> 1. **Data Completeness:** The dataset contains **3,547 entries**. Every single column has **3,547 non-null values**, meaning we have **zero missing data**. This is a great sign of data quality!
> 2. **Data Types:** We have a mix of numerical data (`float64` for revenue, `int64` for hours/sorting) and categorical data (`object` for coffee names and payment types).
> 3. **Memory Usage:** The dataset is lightweight (~305 KB), making it very efficient for processing.

In [3]:
# Check unique values in categorical columns to understand the business domain
print("UNIQUE VALUES")
print("=" * 50)
print(f"\nCoffee Types: {df['coffee_name'].nunique()}")
print(f"Coffee Types: {df['coffee_name'].unique()}")
print(f"\nPayment Types: {df['cash_type'].unique()}")
print(f"\nTime of Day: {df['Time_of_Day'].unique()}")
print(f"\nWeekdays: {df['Weekday'].unique()}")
print(f"\nMonths: {df['Month_name'].unique()}")

UNIQUE VALUES

Coffee Types: 8
Coffee Types: ['Latte' 'Hot Chocolate' 'Americano' 'Americano with Milk' 'Cocoa'
 'Cortado' 'Espresso' 'Cappuccino']

Payment Types: ['card']

Time of Day: ['Morning' 'Afternoon' 'Night']

Weekdays: ['Fri' 'Sat' 'Sun' 'Mon' 'Tue' 'Wed' 'Thu']

Months: ['Mar' 'Apr' 'May' 'Jun' 'Jul' 'Aug' 'Sep' 'Oct' 'Nov' 'Dec' 'Jan' 'Feb']


### 2- Data Cleaning & Preparation
To perform a correct time-series analysis, we need to ensure our `Date` column is recognized as a time object, not just a string of text.

In [4]:
# checking for duplicates to ensure data integrity
print(f'Total Duplicates Row: {df.duplicated().sum()}')

# converting 'Date' from obj to datetime 
df['Date']=pd.to_datetime(df['Date'])

df.head()

Total Duplicates Row: 0


,hour_of_day,cash_type,money,coffee_name,Time_of_Day,Weekday,Month_name,Weekdaysort,Monthsort,Date,Time
0,10,card,38.7,Latte,Morning,Fri,Mar,5,3,2024-03-01,10:15:50.520000
1,12,card,38.7,Hot Chocolate,Afternoon,Fri,Mar,5,3,2024-03-01,12:19:22.539000
2,12,card,38.7,Hot Chocolate,Afternoon,Fri,Mar,5,3,2024-03-01,12:20:18.089000
3,13,card,28.9,Americano,Afternoon,Fri,Mar,5,3,2024-03-01,13:46:33.006000
4,13,card,38.7,Latte,Afternoon,Fri,Mar,5,3,2024-03-01,13:48:14.626000


> #### Cleaning Results:
> * **No Redundancy:** We found **0 duplicates**, so every row represents a unique transaction.
> * **Time-Ready:** The `Date` column is now in `datetime64` format, allowing us to extract trends over weeks, months, or years accurately.

#  Phase 2: Translating Business Questions into Visuals

##  Business Question 1: Weekly Revenue Performance
**Question:** Which day of the week is our "Golden Day" in terms of total revenue?

In [5]:
# Grouping data by Weekday and summing the revenue
weekday_rev = df.groupby('Weekday')['money'].sum().reset_index()

# Visualizing the total revenue per day using a Bar Chart
fig1 = px.bar(weekday_rev, x='Weekday', y='money',
              title='Total Revenue per Day of the Week')

# Ordering the x-axis to show the highest revenue first
fig1.update_layout(title_x=0.5) # to make title in the center
fig1.update_xaxes(categoryorder='total descending')
fig1.show()

> ###  Insight:
> By ranking the days, we can identify which days have the highest traffic. This helps in **staffing decisions**—knowing when we need more baristas on the floor.

##  Business Question 2: Revenue Trend Over Time
**Question:** How is our daily revenue trending across the dataset dates?

In [6]:
df['Date'].dt.year

# Group by date to calculate daily revenue
revTre_df=df.groupby('Date')['money'].sum().reset_index()

# we could use area chart to show the trend of revenue over time
fig=px.line(data_frame=revTre_df,x='Date',y='money', markers=True,title="Daily Revenue Trend")
fig.show()

# Alternative visualization using an area chart
px.area(data_frame=revTre_df,x='Date',y='money',title="Daily Revenue Trend").show()

> ####  Insight:
> Tracking daily revenue helps identify overall growth patterns, seasonal spikes, or specific days that underperformed.

##  Business Question 3: Top Revenue-Generating Products
**Question:** Which coffee type brings in the most total revenue?

In [7]:
cf_type=df.groupby('coffee_name')['money'].sum().reset_index().sort_values(by='money',ascending=False).round()

# Bar chart to show revenue per coffee type
fig=px.bar(data_frame=cf_type,x='coffee_name',y='money',title="Revenue by Coffe Type",color='coffee_name',
           labels={'coffee_name':'Coffee Type','money':'Total Revenue'},text_auto=True)
fig.update_layout(title_x=0.5)
fig.update_xaxes(categoryorder='total descending')
fig.show()

# Treemap for an alternative hierarchical view of revenue
figg=px.treemap(data_frame=cf_type,path=['coffee_name'],values='money',title="Revenue by Coffee Type")
figg.update_layout(title_x=0.5)
figg.show()

> #### Insight:
> Visualizing revenue by product clearly highlights our best-sellers. This is essential for inventory planning and marketing focus.

##  Business Question 4: Payment Method Share
**Question:** What is the percentage share of each payment method (Cash vs. Card)?

In [8]:
# although we see that payment type only has one unique value
# Calculate the count for each payment method
pay_counts=df['cash_type'].value_counts().reset_index()
fig4=px.pie(data_frame=pay_counts,values='count',names='cash_type',hole=0.4,title='Payment Method Share')
fig4.update_layout(title_x=0.5)
fig4.show()

##  Business Question 5: Sales Volume by Hour
**Question:** How does the number of transactions vary by the hour of the day?

In [9]:
# we set the number of bins to 24 to represent each hour of the day
# Histogram to show the distribution of transactions over 24 hours
fig5=px.histogram(data_frame=df,x='hour_of_day',nbins=24,title="Sales Volume by Hour of Day",text_auto=True)
fig5.update_layout(title_x=0.5)
fig5.show()

In [10]:
# another solution: grouping by hour and using Graph Objects
sales_by_hour=df.groupby('hour_of_day').size().sort_index()
fig_5=go.Figure(data=[
    go.Bar(
        x=sales_by_hour.index,
        y=sales_by_hour.values,
        text=[f'{int(val)}' for val in sales_by_hour],
        textposition='outside',
        hovertemplate='<b>Hour:</b> %{x}:00<br><b>Transactions:</b> %{y}<extra></extra>'

    )
])
fig_5.update_layout(
    title={'text': 'Sales Volume by Hour of the Day', 'x': 0.5, 'xanchor': 'center', 'font': {'size': 18}},
    xaxis_title='Hour of Day',
    yaxis_title='Number of Transactions',
    template='plotly_white',
    height=500,
    showlegend=False,
    xaxis=dict(tickmode='linear', tick0=0, dtick=1)
)

fig_5.show()
print(f"Peak sales hour: {sales_by_hour.idxmax()}:00 with {sales_by_hour.max()} transactions")

Peak sales hour: 10:00 with 328 transactions


## Business Question 6: Average Spending by Time of Day
**Question:** How much do customers spend on average during the Morning, Afternoon, and Night?

In [11]:
# Calculate average revenue per time of day
time_of=df.groupby('Time_of_Day')['money'].mean().sort_values().reset_index()
fig6=px.bar(data_frame=time_of,x='Time_of_Day',y='money',title="Average Revenue by Time of Day",color='Time_of_Day',
           labels={'Time_of_Day':'Time of Day','money':'Average Revenue'},text_auto=True)
fig6.update_layout(title_x=0.5)
fig6.show()

In [12]:
# we could use this way to order the time of the day
# by chance the above order based on average revenue is the same as the natural order of time of day, but if it were different we could use this method to set the order explicitly
df['Time_of_Day']=pd.Categorical(df['Time_of_Day'],categories=['Morning','Afternoon','Night'],ordered=True)

## Business Question 7: Average Transaction Value per Weekday
**Question:** Which day of the week sees the highest average transaction value?

In [13]:
# Group by weekday and calculate the mean transaction value
fig7=px.bar(df.groupby('Weekday')['money'].mean().reset_index().round(2),x="Weekday",y='money',title='Average Transactions per WeekDay',text_auto=True,color='Weekday')
fig7.update_layout(title_x=0.5)
fig7.update_xaxes(categoryorder='max descending')
fig7.show()
print('Highest average transaction value: Mon with $31.92')

Highest average transaction value: Mon with $31.92


> ### Insight:
> Even if a day doesn't have the highest total sales, it might have the highest average order value, indicating premium orders or group purchases.

## Business Question 8: Weekday vs. Weekend Performance
**Question:** How does total revenue compare between weekdays and weekends?

In [14]:
df['IS_weekend']=df['Weekday'].apply(lambda x: 'Weekend' if x in [ 'Sat', 'Sun'] else 'Weekday')
weekend_df=df.groupby('IS_weekend')['money'].sum().reset_index()

In [15]:
# Visualize the comparison
fig8=px.bar(data_frame=weekend_df,text_auto=True,x='IS_weekend',y='money',title='Revenue difference between weekends and weekdays',labels={'IS_weekend':'days'})
fig8.update_layout(title_x=0.5)
fig8.show()
print("Revenue comparison:")
for day_type, revenue in weekend_df.iterrows():
    print(f"{day_type}: ${revenue['money']:.2f}")

print("Insights: so the weekdays generate more revenue than the weekends, which could be due to higher foot traffic during the week or special promotions that attract customers on weekdays.\nThis insight can help the coffee shop optimize staffing and marketing strategies to maximize revenue on both weekdays and weekends.")

Revenue comparison:
0: $84176.00
1: $28069.58
Insights: so the weekdays generate more revenue than the weekends, which could be due to higher foot traffic during the week or special promotions that attract customers on weekdays.
This insight can help the coffee shop optimize staffing and marketing strategies to maximize revenue on both weekdays and weekends.


## Business Question 9: Top 5 Revenue-Generating Products
**Question:** Which 5 specific coffee products contribute the most to the overall revenue?

In [16]:
# Get the top 5 coffee products by total revenue
top_coffee=df.groupby('coffee_name')['money'].sum().reset_index().sort_values(by='money',ascending=False).head(5)
fig9=px.bar(top_coffee,x='money',y='coffee_name',orientation='h',title='Top 5 Coffee Revenue',color='coffee_name')
fig9.update_layout(title_x=0.5)
fig9.show()

## Business Question 10: Monthly Revenue Contribution
**Question:** How much does each month contribute to the total revenue?

In [17]:
# Calculate total revenue per month
month_re=df.groupby('Month_name')['money'].sum().reset_index()
fig10=px.bar(month_re,x='Month_name',y='money',title="Total Revenue per Month",color='Month_name',text_auto=True)
fig10.update_xaxes(categoryorder='max descending')
fig10.update_layout(title_x=0.5,showlegend=False)
fig10.show()

## Business Question 11: Coffee Popularity by Time of Day
**Question:** Which coffee type is most popular during the Morning, Afternoon, and Night?

In [19]:
# Group data to see which coffee sells best in each time period
# Adding observed=False to handle the Categorical data type warning
most_popular_coffee = df.groupby(['coffee_name','Time_of_Day'], observed=False)['money'].size().sort_values().reset_index()

# Visualizing the popularity using a grouped bar chart
fig11 = px.bar(most_popular_coffee, x='coffee_name', y='money',
               title="Most Popular Coffee Types by Time of Day",
               color='Time_of_Day', barmode='group',
               labels={'money':'Number of Transactions', 'coffee_name':'Coffee Type'})
fig11.update_layout(title_x=0.5)
fig11.show()

## Business Question 12: Payment Methods per Weekday
**Question:** How does the transaction volume differ between Cash and Card across the weekdays?

In [21]:
# Count transactions per weekday and payment method
z_data = df.groupby(['Weekday', 'cash_type']).size().reset_index(name='count')

# Displaying the comparison
fig12 = px.bar(z_data, 
               x='Weekday', 
               y='count', 
               color='cash_type', 
               barmode='group', 
               text_auto=True,  
               title='Total Transactions: Day vs Payment Method',
               category_orders={'Weekday': ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']})

fig12.update_traces(textfont_color='black') 
fig12.update_layout(title_x=0.5, yaxis_title='Number of Transactions')

fig12.show()

> ### Insight:
> Card transactions strongly dominate every single day of the week. The shop must ensure continuous network stability for its Point of Sale (POS) machines.

## Business Question 13: Spending Distribution per Coffee Type
**Question:** What is the distribution of money spent for each coffee type?

In [22]:
# Box plot to analyze price spread and identify potential outliers
fig13=px.box(df,x='coffee_name',y='money',color='coffee_name',title="Spending Distribution per Coffee Type")
fig13.update_layout(title_x=0.5,showlegend=False)
fig13.show()

# Display the descriptive statistics for deep analysis
print("Summary statistics by coffee type:")
print(df.groupby('coffee_name')['money'].describe())

Summary statistics by coffee type:
                     count       mean       std    min    25%    50%    75%  \
coffee_name                                                                   
Americano            564.0  25.975638  1.676099  23.02  25.96  25.96  25.96   
Americano with Milk  809.0  30.594710  1.883843  27.92  27.92  30.86  30.86   
Cappuccino           486.0  35.883004  1.819232  32.82  35.76  35.76  37.72   
Cocoa                239.0  35.653389  1.233693  32.82  35.76  35.76  35.76   
Cortado              287.0  25.731220  2.093669  23.02  23.02  25.96  27.92   
Espresso             129.0  20.854884  1.974016  18.12  18.12  21.06  23.02   
Hot Chocolate        276.0  35.990797  1.437499  32.82  35.76  35.76  35.76   
Latte                757.0  35.502378  1.821834  32.82  32.82  35.76  35.76   

                      max  
coffee_name                
Americano            28.9  
Americano with Milk  33.8  
Cappuccino           38.7  
Cocoa                38.7  
Cortad

> ### Insight:
> The box plot effectively shows the pricing stability. Most products have a tight price range, indicating standardized pricing without extreme discounts or upcharges.

## Business Question 14: Revenue Contribution by Hour
**Question:** Which hour of the day contributes the most to the total revenue?

In [23]:
# Aggregate total revenue grouped by the hour of the day
hour_rev = df.groupby('hour_of_day')['money'].sum().reset_index()

fig14 = px.bar(hour_rev, x='hour_of_day', y='money',text_auto=True,
               title='Total Revenue Contribution by Hour')
fig14.update_traces(textfont_color='black', textposition='outside') 
fig14.update_layout(title_x=0.5)
fig14.show()

print('Peak revenue hour: 10:00 with $10,198.52')

Peak revenue hour: 10:00 with $10,198.52


## Business Question 15: Revenue Share per Coffee Type
**Question:** What is the overall percentage share of each coffee type in the total revenue?

In [24]:
# Pie chart to represent the revenue market share of each product
fig15 = px.pie(df, values='money', names='coffee_name', hole=0.4,
               title='Overall Revenue Share per Coffee Type')

fig15.update_traces(textinfo='percent+label') 
fig15.update_layout(title_x=0.5)
fig15.update_traces(textfont_color='black',textposition='inside')
fig15.show()

## Business Question 16: Hourly Sales Trend
**Question:** Is there a clear peak sales time in the day based on transaction count?

In [25]:
# Count the number of transactions per hour to observe the volume flow
hourly_trend = df.groupby('hour_of_day').size().reset_index(name='sales_count')

fig16 = px.line(hourly_trend, x='hour_of_day', y='sales_count', markers=True,
                title='Hourly Sales Trend (Peak Times)')

fig16.update_layout(title_x=0.5)
fig16.show()

## Business Question 17: Total Coffee Sold by Time of Day
**Question:** Which period of the day (Morning, Afternoon, Night) records the highest number of coffees sold?

In [26]:
# Count total occurrences per broad time period
time_counts = df['Time_of_Day'].value_counts().reset_index()

fig18 = px.bar(time_counts, x='Time_of_Day', y='count', color='Time_of_Day',
               title='Total Quantity of Coffee Sold by Time Period')

fig18.update_layout(title_x=0.5, showlegend=False)
fig18.show()

> ### Insight:
> Although the morning has the single busiest hour, the "Afternoon" shift collectively handles the highest volume of cups sold.

## Business Question 18: Cumulative Revenue Trend
**Question:** What is the cumulative revenue growth over the entire dataset timeline?

In [28]:
# Sort the dataframe by Date chronologically to ensure accurate cumulative sum
df_sorted = df.sort_values('Date')
daily_cum = df_sorted.groupby('Date')['money'].sum().cumsum().reset_index()

# Plot the running total
fig20 = px.line(daily_cum, x='Date', y='money', 
                title='Cumulative Revenue Growth Over Time')

fig20.update_layout(title_x=0.5)
fig20.show()

> ### Insight:
> The cumulative line (`cumsum`) displays a healthy, linear growth. The absence of flat periods means the business is generating consistent daily income without significant interruptions.